packages/core/src/compiler/styleTypeChecker/styleTypeChecker.ts

```diff
@@ -0,0 +1,102 @@
+import { Shape, ShapeType } from "../../shapes/Shapes";
+import * as ad from "../../types/ad";
+import { ValueShapeT, ValueType } from "../../types/types";
+import { ArgVal, Value } from "../../types/value";
+
+export const checkType = (
+  expectedType: ValueShapeT,
+  arg: ArgVal<ad.Num>
+): Shape<ad.Num> | Value<ad.Num>["contents"] | undefined => {
+  if (expectedType.tag === "UnionT") {
+    for (const t of expectedType.types) {
+      const result = checkType(t, arg);
+      if (result !== undefined) return result;
+    }
+  } else if (expectedType.tag === "ShapeT") {
+    const shape = checkShape(expectedType.type, arg);
+    if (shape !== undefined) return shape;
+  } else {
+    const value = checkArgValAgainstValueType(expectedType.type, arg);
+    if (value !== undefined) return value;
+  }
+  return undefined;
+};
+
+export const checkShape = (
+  expectedType: ShapeType | "AnyShape",
+  arg: ArgVal<ad.Num>
+): Shape<ad.Num> | undefined => {
+  if (expectedType === "AnyShape") {
+    if (arg.tag === "ShapeVal") {
+      return arg.contents;
+    }
+  } else {
+    if (arg.tag === "ShapeVal" && arg.contents.shapeType === expectedType) {
+      return arg.contents;
+    }
+  }
+  return undefined;
+};
+
+export const checkArgValAgainstValueType = (
+  expected: ValueType,
+  given: ArgVal<ad.Num>
+): Value<ad.Num>["contents"] | undefined => {
+  if (given.tag === "ShapeVal") {
+    return undefined;
+  } else {
+    return checkValueAgainstValueType(expected, given.contents);
+  }
+};
+
+export const checkValueAgainstValueType = (
+  expected: ValueType,
+  given: Value<ad.Num>
+): Value<ad.Num>["contents"] | undefined => {
+  const { tag, contents } = given;
+  if (
```

Comment:
Use `switch` to just fall through if there's no checking to do.

Reply:
I want this function to return `undefined` if the check fails. The idea is, `checkValueAgainstValueType(type, value)` will return `value.contents` if `value` is consistent with `type`, and `undefined` otherwise. I guess there isn't really much difference in this case?

Reply:
ahh I see the two-layer `if`s. Okay with keeping this design then because I don't see a switch improving the it more.

Reply:
Sounds good